scripts: Migrate scan rule functionality from core

ricekot opened the pull request:
## Overview
Migrate scripts scanning infrastructure from core to add-ons.

## Related Issues
- First step towards https://github.com/zaproxy/zaproxy/issues/7105.
- See also https://github.com/zaproxy/zaproxy/pull/8265.

## Checklist
- [ ] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [ ] Check code coverage
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

thc202 commented on addOns/jruby/jruby.gradle.kts:
```diff
@@ -9,6 +9,13 @@ zapAddOn {
+        dependencies {
+            addOns {
+                register("scripts") {
+                    version.set(">=45")
```
Might be a good time to start using semver in the scripts add-on.

thc202 commented on addOns/scripts/CHANGELOG.md:
```diff
@@ -4,7 +4,8 @@ All notable changes to this add-on will be documented in this file.
 ## Unreleased
-
+### Added
+- The scan rule functionality of scripts was moved from the ZAP core to this add-on.
```
Reference issue 7105?

thc202 commented on addOns/scripts/src/main/java/org/zaproxy/zap/extension/scripts/scanrules/PassiveScript.java:
```diff
@@ -0,0 +1,46 @@
+     * @param historyType the history type of the message to be scanned.
+     * @return {@code true} if the scanner applies to the given history type, {@code false}
+     *     otherwise.
+     * @since 2.7.0
```
Technically these are all new classes and interfaces, the `since`s could be dropped.

thc202 commented:
Didn't mean to approve…

Sometimes I get:
```
[SwingWorker-pool-3-thread-1] ERROR org.zaproxy.zap.control.AddOnInstaller - Failed to uninstall pscanrule: org.zaproxy.zap.extension.scripts.scanrules.ScriptsPassiveScanner
```
but still not sure how to reproduce that.

ricekot commented:
Addressed review comments. For the record, the error in the above comment was happening because the rule bundled with the add-on was not being installed in 2.14 (related: https://github.com/zaproxy/zaproxy/pull/8274). The fix was to install the rule in `ExtensionScriptsUI#postInit`.

thc202 commented on addOns/scripts/src/main/java/org/zaproxy/zap/extension/scripts/scanrules/ScriptsActiveScanner.java:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    /**
+     * @deprecated (2.9.0) Use {@link #newAlert()} to build and {@link AlertBuilder#raise() raise}
```
The versions in these could be removed as well (or updated to 45.0.0?).

thc202 commented on addOns/scripts/src/test/java/org/zaproxy/zap/extension/scripts/scanrules/ScriptsPassiveScannerUnitTest.java:
```diff
@@ -0,0 +1,330 @@
+import org.zaproxy.zap.extension.scripts.ExtensionScriptsUI;
+import org.zaproxy.zap.testutils.TestUtils;
+
+/** Unit test for {@link org.zaproxy.zap.extension.pscan.scanner.ScriptsPassiveScanner}. */
```
Remove package?

thc202 commented on addOns/scripts/src/main/java/org/zaproxy/zap/extension/scripts/scanrules/ScriptsActiveScanner.java:
```diff
@@ -0,0 +1,370 @@
+
+    @Override
+    public String getDescription() {
+        return "";
```
Kudos for removing the N/As :)

thc202 commented:
Thank you!
